### PR TITLE
fix: revoke session may be ignored due to server context cancellation in advance

### PR DESCRIPTION
**Pull request by jaime0815**
issue: #31219
pr: #31220

### CODE REVIEW

File: internal/distributed/querynode/service.go [hunk 1]
```diff
@@ -241,19 +241,22 @@ func (s *Server) Stop() (err error) {
...
 	}
 
-	s.cancel()
 	if s.grpcServer != nil {
 		utils.GracefulStopGRPCServer(s.grpcServer)
 	}
 	s.wg.Wait()
+
+	s.cancel()
 	return nil
```
Comment (weiliu1031) [inline internal/distributed/querynode/service.go:260]: should call `cancel` before `wg.Wait()`?

**Comment by weiliu1031** [comment]
the `wg` in `distributed/*/service.go` is used to wait grpc server to exit, so `ctx.Cannel()` happens after `wg.Wait()`.

File: internal/distributed/querynode/service.go [hunk 2]
```diff
@@ -241,19 +241,22 @@ func (s *Server) Stop() (err error) {
...
 		defer s.etcdCli.Close()
 	}
 
-	s.cancel()
 	if s.grpcServer != nil {
 		utils.GracefulStopGRPCServer(s.grpcServer)
 	}
 	s.wg.Wait()
+
+	s.cancel()
```
Comment (czs007) [inline internal/distributed/querynode/service.go:259]: should be ahead of s.wg.Wait()?

File: internal/distributed/datacoord/service.go [hunk 3]
```diff
@@ -237,14 +235,16 @@ func (s *Server) Stop() (err error) {
...
+	logger.Info("internal server[dataCoord] start to stop")
 	err = s.dataCoord.Stop()
 	if err != nil {
+		log.Error("failed to close dataCoord", zap.Error(err))
 		return err
 	}
 
-	s.wg.Wait()
-
+	s.cancel()
```
Comment (czs007) [inline internal/distributed/datacoord/service.go:247]: should be ahead of s.wg.Wait() ?

File: internal/distributed/datanode/service.go [hunk 4]
```diff
@@ -201,24 +201,27 @@ func (s *Server) Run() error {
...
 
+	logger.Info("internal server[datanode] start to stop")
 	err = s.datanode.Stop()
 	if err != nil {
+		log.Error("failed to close datanode", zap.Error(err))
 		return err
 	}
+
 	s.wg.Wait()
+	s.cancel()
```
Comment (czs007) [inline internal/distributed/datanode/service.go:224]: ditto

File: internal/distributed/querynode/service.go (hunk 2 above)
Comment (jaime0815) [reply internal/distributed/querynode/service.go:259]: wg.Wait() is only used to wait grpc server to exit

**Comment by xiaofan-luan** [comment]
generally speaking, put cancel before s.wg.wait() could be safer.
Agreed we should do this change!

**Comment by jaime0815** [comment]
> generally speaking, put cancel before s.wg.wait() could be safer. Agreed we should do this change!
we can move the partial cancel() method before s.wg.wait(), cancel will lead to context done,  server may exit in advance is not expected

**Comment by jaime0815** [comment]
image
some goroutines in s.dataCoord will exit before executing s.dataCoord.Stop() if we move s.cancel() in front of s.wg.Wait()

**Comment by yanliang567** [comment]
rerun the failed ut tests

**Comment by yanliang567** [comment]
manual pass as one occaisonal pass

**Pull request merged**